src/matrix/kernelmatrix.jl

```diff
@@ -8,74 +8,82 @@ kernelmatrix!
+    κ::Kernel,
+    X::AbstractMatrix;
+    obsdim::Int = defaultobs
+)
+    @assert obsdim ∈ [1, 2] "obsdim should be 1 or 2 (see docs of `kernelmatrix`))"
+    if obsdim == 1
+        kernelmatrix!(K, κ, ColVecs(X'))
+    else
+        kernelmatrix!(K, κ, ColVecs(X))
+    end
 end
 
-kernelmatrix!(K::AbstractMatrix, κ::TransformedKernel, X::AbstractMatrix; obsdim::Int = defaultobs) =
-        kernelmatrix!(K, kernel(κ), apply(κ.transform, X, obsdim = obsdim), obsdim = obsdim)
+function kernelmatrix!(
+    K::AbstractMatrix,
+    κ::Kernel,
+    X::AbstractVector
+    )
+    if !check_dims(K, X, X)
+        throw(DimensionMismatch("Dimensions of the target array K $(size(K)) are not consistent with X $(size(X))"))
+    end
+    K .= κ.(X, X')
+end
 
 function kernelmatrix!(
     K::AbstractMatrix,
-    kernel::Kernel,
-    X::AbstractVector,
-    Y::AbstractVector;
+    κ::SimpleKernel,
+    X::AbstractMatrix,
+    Y::AbstractMatrix;
     obsdim::Int = defaultobs,
 )
-    return kernelmatrix!(K, kernel, reshape(X, 1, :), reshape(Y, 1, :); obsdim = 2)
+    @assert obsdim ∈ [1, 2] "obsdim should be 1 or 2 (see docs of kernelmatrix))"
+    if !check_dims(K, X, Y, feature_dim(obsdim), obsdim)
+        throw(DimensionMismatch("Dimensions $(size(K)) of the target array K are not consistent with X ($(size(X))) and Y ($(size(Y)))"))
+    end
+    map!(x -> kappa(κ, x), K, pairwise(metric(κ), X, Y, dims = obsdim))
 end
 
 function kernelmatrix!(
-        K::AbstractMatrix,
-        κ::Kernel,
-        X::AbstractMatrix,
-        Y::AbstractMatrix;
-        obsdim::Int = defaultobs
-        )
-        @assert obsdim ∈ [1,2] "obsdim should be 1 or 2 (see docs of kernelmatrix))"
-        if !check_dims(K,X,Y,feature_dim(obsdim),obsdim)
-            throw(DimensionMismatch("Dimensions $(size(K)) of the target array K are not consistent with X ($(size(X))) and Y ($(size(Y)))"))
-        end
-        map!(x->kappa(κ,x),K,pairwise(metric(κ),X,Y,dims=obsdim))
-end
-
-kernelmatrix!(K::AbstractMatrix, κ::TransformedKernel, X::AbstractMatrix, Y::AbstractMatrix; obsdim::Int = defaultobs) =
-        kernelmatrix!(K, kernel(κ), apply(κ.transform, X, obsdim = obsdim), apply(κ.transform, Y, obsdim = obsdim), obsdim = obsdim)
-
-## Apply kernel on two reals ##
-function _kernel(κ::Kernel, x::Real, y::Real)
-    _kernel(κ, [x], [y])
-end
-
-## Apply kernel on two vectors ##
-function _kernel(
-        κ::Kernel,
-        x::AbstractVector,
-        y::AbstractVector;
-        obsdim::Int = defaultobs
-    )
-    @assert length(x) == length(y) "x and y don't have the same dimension!"
-    kappa(κ, evaluate(metric(κ),x,y))
+    K::AbstractMatrix,
+    κ::Kernel,
+    X::AbstractMatrix,
+    Y::AbstractMatrix;
+    obsdim::Int = defaultobs
+)
+    @assert obsdim ∈ [1, 2] "obsdim should be 1 or 2 (see docs of `kernelmatrix`))"
+    if obsdim == 1
+        kernelmatrix!(K, κ, ColVecs(X'), ColVecs(Y'))
+    else
+        kernelmatrix!(K, κ, ColVecs(X), ColVecs(Y))
+    end
```

Comment:
Similarly, this could be written as
```suggestion
    kernelmatrix!(K, κ, vec_of_vecs(X; obsdim = obsdim), vec_of_vecs(Y; obsdim = obsdim))
```